modules/asset-registry/src/lib.rs

```diff
@@ -51,17 +64,74 @@ pub mod module {
 
 	#[pallet::config]
 	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Currency type for withdraw and balance storage.
 		type Currency: Currency<Self::AccountId>;
+
+		/// Evm Bridge for getting info of contracts from the EVM.
 		type EVMBridge: EVMBridge<Self::AccountId, BalanceOf<Self>>;
+
+		/// Required origin for registering asset.
+		type RegisterOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Weight information for the extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[derive(Clone, Eq, PartialEq, RuntimeDebug, Encode, Decode, TypeInfo)]
+	pub struct AssetMetadata<Balance> {
+		pub name: Vec<u8>,
+		pub symbol: Vec<u8>,
+		pub decimals: u8,
+		pub minimal_balance: Balance,
 	}
 
-	/// Error for evm accounts module.
 	#[pallet::error]
 	pub enum Error<T> {
+		/// The given location could not be used (e.g. because it cannot be expressed in the
+		/// desired version of XCM).
+		BadLocation,
+		/// AssetMetadata already existed
+		AssetMetadataExisted,
+		/// AssetMetadata not exists
+		AssetMetadataNotExists,
 		/// CurrencyId existed
 		CurrencyIdExisted,
 	}
 
+	#[pallet::event]
+	#[pallet::generate_deposit(fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Registered foreign asset. \[ForeignAssetId, AssetMetadata\]
+		RegisteredForeignAsset(ForeignAssetId, MultiLocation, AssetMetadata<BalanceOf<T>>),
+		/// Updated foreign asset. \[AssetMetadata\]
+		UpdatedForeignAsset(MultiLocation, AssetMetadata<BalanceOf<T>>),
+	}
+
+	/// Next available Foreign AssetId ID.
+	///
+	/// NextForeignAssetId: ForeignAssetId
+	#[pallet::storage]
+	#[pallet::getter(fn next_foreign_asset_id)]
+	pub type NextForeignAssetId<T: Config> = StorageValue<_, ForeignAssetId, ValueQuery>;
+
+	/// The storages for MultiLocations.
+	///
+	/// MultiLocations: map ForeignAssetId => Option<MultiLocation>
+	#[pallet::storage]
+	#[pallet::getter(fn multi_locations)]
+	pub type MultiLocations<T: Config> = StorageMap<_, Twox64Concat, ForeignAssetId, MultiLocation, OptionQuery>;
+
+	/// The storages for AssetMetadatas.
+	///
+	/// AssetMetadatas: map MultiLocation => Option<AssetMetadata>
+	#[pallet::storage]
+	#[pallet::getter(fn asset_metadatas)]
+	pub type AssetMetadatas<T: Config> =
+		StorageMap<_, Twox64Concat, MultiLocation, AssetMetadata<BalanceOf<T>>, OptionQuery>;
```

Comment:
use `ForeignAssetId` as key can reduce migration required when upgrading MultiLocation in later xcm version